File: algorithms/11_golang/geometry/graham_scan.go
package geometry
import "sort"
type Point struct{ X, Y int }
func cross(o, a, b Point) int { return (a.X-o.X)*(b.Y-o.Y) - (a.Y-o.Y)*(b.X-o.X) }
func GrahamScan(pts []Point) []Point {
    sort.Slice(pts, func(i, j int) bool {
        if pts[i].X == pts[j].X { return pts[i].Y < pts[j].Y }
        return pts[i].X < pts[j].X
    })
    hull := []Point{}
    for _, p := range pts {
        for len(hull) >= 2 && cross(hull[len(hull)-2], hull[len(hull)-1], p) <= 0 {
            hull = hull[:len(hull)-1]
        }
        hull = append(hull, p)
    }
    return hull
}

File: algorithms/03_cpp/geometry/graham_scan.cpp
#include <vector>
#include <algorithm>
#include <cassert>
#include <iostream>
struct Point { long long x, y; };
long long cross(Point o, Point a, Point b) {
    return (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x);
}
std::vector<Point> convexHull(std::vector<Point> pts) {
    std::sort(pts.begin(), pts.end(), [](Point a, Point b) {
        return a.x < b.x || (a.x == b.x && a.y < b.y);
    });
    std::vector<Point> hull;
    for (auto p : pts) {
        while (hull.size() >= 2 && cross(hull[hull.size() - 2], hull.back(), p) <= 0) {
            hull.pop_back();
        }
        hull.push_back(p);
    }
    return hull;
}
int main() {
    auto h = convexHull({{0, 0}, {1, 1}, {2, 2}, {0, 2}, {2, 0}});
    assert(!h.empty());
    std::cout << "C++ Graham Scan verified.\n";
}
